File: Programmers/jsh/9주차(힙)/더 맵게.py
import heapq #우선순위 큐
def solution(scoville, K):
    answer=0
    heapq.heapify(scoville) #리스트를 최소힙 형태로 변환
    while scoville[0]<K: #top에 있는 수가 k이상이면 모든원소가 k이상이므로
        if len(scoville)==1:
            return -1
        #두개를 pop하여 계산후 다시 push해줌
        heapq.heappush(scoville,heapq.heappop(scoville)+heapq.heappop(scoville)*2)
        answer+=1
    return answer

File: Programmers/Segun/9주차 - 힙/더 맵게.cpp
//heap구조를 사용하는 priority_queue 를 사용해서 문제 해결
//우선순위큐를 오름차순으로 정렬되게 기준을 설정해주고 앞에 두 값을 pop하고 그 두값을 섞은 값을 다시 push해줌

#include <string>
#include <vector>
#include<queue>

using namespace std;

int solution(vector<int> scoville, int K) {
    int answer = 0;
    priority_queue<int,vector<int>,greater<int>> pq(scoville.begin(),scoville.end());

    while(pq.top()<K){
        if(pq.size()==1){return -1;}
        int x=pq.top();pq.pop();
        int y=pq.top();pq.pop();
        pq.push(x+y*2);
        answer++;
    }
    return answer;
}
